### PR TITLE
feat(core): add support for template output

### PR DESCRIPTION
**Pull request by remyleone**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request.
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Relates OR Closes #0000

Release note for [CHANGELOG](https://github.com/scaleway/scaleway-cli/blob/v2/CHANGELOG.md):

```release-note

```

### CODE REVIEW

File: internal/core/printer.go [hunk 1]
```diff
@@ -225,3 +263,36 @@ func (p *Printer) printYAML(data interface{}) error {
...
+	switch dataType.Kind() {
+	// If we have a slice of value, we apply the template for each item
+	case reflect.Slice:
+		elemType := dataType.Elem()
+		for i := 0; i < dataValue.Len(); i++ {
+			elemValue := dataValue.Index(i)
+			err := p.template.Execute(writer, elemValue)
+			if err != nil {
+				return &CliError{
+					Err: fmt.Errorf("templating error. Acceptable values are %s", gofields.ListFields(elemType)),
```
Comment (jerome-quere) [inline internal/core/printer.go:285]: strings.Join(",", gofields.ListFields(elemType))

File: internal/core/autocomplete.go [hunk 2]
```diff
@@ -58,6 +58,13 @@ type FlagSpec struct {
 }
 
 func (node *AutoCompleteNode) addGlobalFlags() {
+	printerType := []string{
```
Comment (kindermoumoute) [inline internal/core/autocomplete.go:61]: nit: `printerTypes`

File: internal/core/printer.go [hunk 3]
```diff
@@ -30,6 +32,9 @@ const (
 
 	// Option to enable pretty output on json printer.
 	PrinterOptJSONPretty = "pretty"
+
+	// PrinterTypeTemplate defines a go template to use to format output.
+	PrinterTypeTemplate = PrinterType("template")
```
Comment (kindermoumoute) [inline internal/core/printer.go:37]: nit: move after line 31

File: internal/core/printer.go [hunk 4]
```diff
@@ -225,3 +263,35 @@ func (p *Printer) printYAML(data interface{}) error {
 
 	return encoder.Encode(data)
 }
+
+func (p *Printer) printTemplate(data interface{}) error {
+	writer := p.stdout
+	_, isError := data.(error)
+	if isError {
```
Comment (kindermoumoute) [inline internal/core/printer.go:270]: nit
```suggestion
	if _, isError := data.(error); isError {
```

File: internal/core/printer.go [hunk 5]
```diff
@@ -225,3 +263,35 @@ func (p *Printer) printYAML(data interface{}) error {
...
+		for i := 0; i < dataValue.Len(); i++ {
+			elemValue := dataValue.Index(i)
+			err := p.template.Execute(writer, elemValue)
+			if err != nil {
+				return p.printHuman(&CliError{
+					Err:  fmt.Errorf("templating error"),
+					Hint: fmt.Sprintf("Acceptable values are:\n  - %s", strings.Join(gofields.ListFields(elemValue.Type()), "\n  - ")),
+				}, nil)
+			}
+			_, _ = writer.Write([]byte{'\n'})
```
Comment (kindermoumoute) [inline internal/core/printer.go:287]: Why not checking error here? Or can you comment why we don't need to?

File: internal/core/printer.go [hunk 6]
```diff
@@ -225,3 +263,35 @@ func (p *Printer) printYAML(data interface{}) error {
...
+		}
+	default:
+		err := p.template.Execute(writer, data)
+		if err != nil {
+			return err
+		}
+		_, _ = writer.Write([]byte{'\n'})
```
Comment (kindermoumoute) [inline internal/core/printer.go:294]: same as above

File: internal/core/testdata/test-core-printer-human-list-with-options-unknown-column.golden [hunk 7]
```diff
@@ -1,9 +1,6 @@
 🎲🎲🎲 EXIT CODE: 0 🎲🎲🎲
 🟥🟥🟥 STDERR️️ 🟥🟥🟥️
-Unknown field 'Unknown' in output options
-
-Hint:
-Valid fields are: ID, Name
```
Comment (kindermoumoute) [inline internal/core/testdata/test-core-printer-human-list-with-options-unknown-column.golden:6]: why is this removed?

File: internal/core/testdata/test-template-printer-template-list-with-options-unknown-column.golden [hunk 8]
```diff
@@ -0,0 +1,19 @@
+🎲🎲🎲 EXIT CODE: 0 🎲🎲🎲
+🟥🟥🟥 STDERR️️ 🟥🟥🟥️
+Templating error
+
+Hint:
+Acceptable values are:
+  - ID
+  - Name
+🟩🟩🟩 JSON STDOUT 🟩🟩🟩
+[
```
Comment (kindermoumoute) [inline internal/core/testdata/test-template-printer-template-list-with-options-unknown-column.golden:10]: It's confusing to have JSON in goldens for such test

File: internal/core/testdata/test-template-printer-template-list-with-options.golden [hunk 9]
```diff
@@ -0,0 +1,15 @@
+🎲🎲🎲 EXIT CODE: 0 🎲🎲🎲
+🟩🟩🟩 STDOUT️ 🟩🟩🟩️
+David Copperfield <-> 111111111-111111111
```
Comment (kindermoumoute) [inline internal/core/testdata/test-template-printer-template-list-with-options.golden:3]: Now I kinda want to see what was the command from the golden file 🤔

**Review by kindermoumoute — CHANGES_REQUESTED** [review]

File: internal/core/testdata/test-template-printer-template-list-with-options-unknown-column.golden (hunk 8 above)
Comment (remyleone) [reply internal/core/testdata/test-template-printer-template-list-with-options-unknown-column.golden:10]: I don't think we can avoid it. @jerome-quere

File: internal/core/testdata/test-template-printer-template-list-with-options.golden (hunk 9 above)
Comment (remyleone) [reply internal/core/testdata/test-template-printer-template-list-with-options.golden:3]: In this case: `scw list -o template={{ .Name }} <-> {{ .ID }}`
If you want to open a ticket to include in the golden the command feel free to do so :)

**Review by kindermoumoute — APPROVED** [review]
LGTM
